src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -647,18 +673,51 @@ object LazyList extends LazyListFactory[LazyList] {
       this
     }
 
-    protected def tailDefined: Boolean = tlEvaluated
-    protected def headDefined: Boolean = hdEvaluated
+    protected[immutable] def tailDefined: Boolean = tlEvaluated
+    protected[immutable] def headDefined: Boolean = hdEvaluated
   }
 
+  @SerialVersionUID(3L)
+  final class Suspended[A](var eval: () => LazyList[A]) extends LazyList[A] {
+    private[this] var evaluated: Boolean = false
+    lazy val next: LazyList[A] = {
+      evaluated = true
+      var res = eval()
+      eval = null
+      while (res.isInstanceOf[Suspended[_]]) {
+        // skip through multiple suspensions in a loop rather than using the stack
+        res = res.asInstanceOf[Suspended[A]].next
+      }
+      res
+    }
+
+    override def isEmpty: Boolean = next.isEmpty
+    override def head: A = next.head
+    override def tail: LazyList[A] = next.tail
+    def force: this.type = { next.force ; this }
+    override def knownSize: Int = if (evaluated) next.knownSize else -1
+    protected[immutable] def headDefined: Boolean = evaluated && next.headDefined
+    protected[immutable] def tailDefined: Boolean = evaluated && next.tailDefined
+  }
+
+  /** Create a [[LazyList]] which defers its construction.
+    *
+    * This allows for the creation of [[LazyList]]s which do not know whether
+    * they are empty without doing some computation.
+    *
+    * @param eval a closure which will create another [[LazyList]]
+    */
+  def suspend[A](eval: => LazyList[A]): LazyList[A] =
```

Comment:
I'm not sure how I feel about the name `suspend`, both internally and externally, but especially as an externally visible API. How do you feel about the names `defer` and `Deferred`?

Reply:
I'm okay with `defer`. Since we had the `Deferrer` class, I didn't want to overload the word.